instalation and imports

In [ ]:
!rm -rf ~/.cache/huggingface

In [ ]:
!ls -R ~/.cache/huggingface

ls: cannot access '/root/.cache/huggingface': No such file or directory


In [ ]:
!du -sh ~/.cache/huggingface/*


du: cannot access '/root/.cache/huggingface/*': No such file or directory


In [ ]:
# To programowo:
import os
os.kill(os.getpid(), 9)


In [ ]:
!pip install datasets librosa numpy pandas tqdm tensorflow

In [2]:
import tensorflow as tf
from tensorflow.keras.layers import (
    Input, Conv1D, MaxPooling1D, BatchNormalization,
    Dropout, Flatten, Dense, Bidirectional, LSTM, Average
)
from tensorflow.keras.models import Model


2025-06-01 20:12:14.261702: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1748808734.665299      35 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1748808734.785181      35 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


In [ ]:
!pip show huggingface_hub

Name: huggingface-hub
Version: 0.31.1
Summary: Client library to download and publish models, datasets and other repos on the huggingface.co hub
Home-page: https://github.com/huggingface/huggingface_hub
Author: Hugging Face, Inc.
Author-email: julien@huggingface.co
License: Apache
Location: /usr/local/lib/python3.11/dist-packages
Requires: filelock, fsspec, hf-xet, packaging, pyyaml, requests, tqdm, typing-extensions
Required-by: accelerate, datasets, diffusers, peft, sentence-transformers, timm, tokenizers, torchtune, transformers


In [ ]:
from huggingface_hub import HfApi
import os

os.environ["HF_TOKEN"] = "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"


In [2]:
from huggingface_hub import login

login(token="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")

In [3]:
!pip install -U datasets fsspec

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 193.6/193.6 kB 4.9 MB/s eta 0:00:00a 0:00:01
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2025.3.2
    Uninstalling fsspec-2025.3.2:
      Successfully uninstalled fsspec-2025.3.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
cesium 0.12.4 requires numpy<3.0,>=2.0, but you have numpy 1.26.4 which is incompatible.
bigframes 1.42.0 requires rich<14,>=12.4.4, but you have rich 14.0.0 which is incompatible.
torch 2.6.0+cu124 requires nvidia-cublas-cu12==12.4.5.8; platform_system == "Linux" and platform_machine == "x86_64", but you have nvidia-cublas-cu12 12.9.0.13 which is incompatible.
torch 2.6.0+cu124 requires nvidia-cudnn-cu12==9.1.0.70; platform_system == "Linux" and platform_machine == "x86_64", but you have nvidia-cudnn-cu12 9.3.0.75 which is incompatible.
torch 2.6.0+cu124 requires nvid

In [4]:
from datasets import load_dataset
dataset = load_dataset("marbar16/podcastv2", trust_remote_code=True)

podcastv2.py:   0%|          | 0.00/4.10k [00:00<?, ?B/s]

train3.tar.gz:   0%|          | 0.00/6.06G [00:00<?, ?B/s]

test3.tar.gz:   0%|          | 0.00/760M [00:00<?, ?B/s]

train_metadata_v3.csv:   0%|          | 0.00/10.9M [00:00<?, ?B/s]

test_metadata_v3.csv:   0%|          | 0.00/1.33M [00:00<?, ?B/s]

Generating train split: 0 examples [00:00, ? examples/s]

Generating test split: 0 examples [00:00, ? examples/s]

In [24]:
import tensorflow as tf
from tensorflow.keras.layers import (
    Input, Conv1D, MaxPooling1D, BatchNormalization,
    Dropout, Flatten, Dense, Bidirectional, LSTM, Average
)
from tensorflow.keras.models import Model


In [5]:
dataset = dataset.rename_column("category", "label")
dataset

DatasetDict({
    train: Dataset({
        features: ['filename', 'audio', 'path', 'label', 'arousal', 'valence', 'dominance', 'emotion_secondary', 'transcript', 'speaker_id', 'gender'],
        num_rows: 40000
    })
    test: Dataset({
        features: ['filename', 'audio', 'path', 'label', 'arousal', 'valence', 'dominance', 'emotion_secondary', 'transcript', 'speaker_id', 'gender'],
        num_rows: 5000
    })
})

data prep

In [ ]:
import librosa
import numpy as np
import pandas as pd
from tqdm import tqdm
from datasets import load_dataset  # or however you load your dataset

# Load dataset (replace with your loading method)
# dataset = load_dataset("path_to_your_dataset")["train"]  # adjust split if needed

# Parameters
sr_expected = 16000
top_db = 30  # silence threshold in dB
min_non_silent_ratio = 0.6
min_avg_rmse = 0.01
min_avg_zcr = 0.01  # optional lower threshold
max_avg_zcr = 0.3   # optional upper threshold

results = []

for sample in tqdm(dataset["train"], desc="Analyzing audio samples"):

    audio_array = sample["audio"]["array"]
    sr = sample["audio"]["sampling_rate"]

    if sr != sr_expected:
        # Resample if necessary
        audio_array = librosa.resample(audio_array, orig_sr=sr, target_sr=sr_expected)
        sr = sr_expected

    # 1. Silence Detection
    intervals = librosa.effects.split(audio_array, top_db=top_db)
    non_silent_duration = sum(end - start for start, end in intervals)
    total_duration = len(audio_array)
    non_silent_ratio = non_silent_duration / total_duration

    # 2. ZCR (Zero-Crossing Rate)
    zcr = librosa.feature.zero_crossing_rate(audio_array)[0]
    avg_zcr = np.mean(zcr)

    # 3. RMSE (Energy)
    rmse = librosa.feature.rms(y=audio_array)[0]
    avg_rmse = np.mean(rmse)

    # 4. Decide if usable
    usable = (
        non_silent_ratio >= min_non_silent_ratio
        and avg_rmse >= min_avg_rmse
        and min_avg_zcr <= avg_zcr <= max_avg_zcr
    )

    results.append({
        "filename": sample["filename"],
        "non_silent_ratio": round(non_silent_ratio, 3),
        "avg_zcr": round(avg_zcr, 4),
        "avg_rmse": round(avg_rmse, 4),
        "usable": usable,
        "label": sample["label"],
        "transcript": sample["transcript"]
    })

# Convert results to DataFrame
df = pd.DataFrame(results)

# Save for further inspection
df.to_csv("audio_quality_check.csv", index=False)

# Summary
usable_count = df["usable"].sum()
print(f"\nUsable files: {usable_count}/{len(df)} ({100 * usable_count / len(df):.2f}%)")


In [7]:
import numpy as np
import librosa
from datasets import Dataset
from tqdm import tqdm
from librosa.util import fix_length
# Constants
SAMPLE_RATE = 16000
CHUNK_DURATION = 2.5
CHUNK_SIZE = int(SAMPLE_RATE * CHUNK_DURATION)
HOP_SIZE = CHUNK_SIZE // 2

FRAME_LENGTH = 2048
HOP_LENGTH = 512
MAX_LENGTH = 10  # seconds
MAX_SAMPLES = SAMPLE_RATE * MAX_LENGTH

def extract_features(y, sr):
    # zcr = librosa.feature.zero_crossing_rate(y, frame_length=FRAME_LENGTH, hop_length=HOP_LENGTH)
    rmse = librosa.feature.rms(y=y, frame_length=FRAME_LENGTH, hop_length=HOP_LENGTH)
    mfcc = librosa.feature.mfcc(y=y, sr=sr, n_mfcc=20, hop_length=HOP_LENGTH)
    chroma = librosa.feature.chroma_stft(y=y, sr=sr, hop_length=HOP_LENGTH)

    return np.concatenate([
        # zcr.mean(axis=1), zcr.std(axis=1),
        rmse.mean(axis=1), rmse.std(axis=1),
        mfcc.mean(axis=1), mfcc.std(axis=1),
        chroma.mean(axis=1), chroma.std(axis=1)
    ])

def segment_and_extract(audio_array, sr):
    features = []

    for start in range(0, len(audio_array) - CHUNK_SIZE + 1, HOP_SIZE):
        chunk = audio_array[start:start + CHUNK_SIZE]
        if np.abs(chunk).mean() < 1e-4:  # too silent
            continue

        feats = extract_features(chunk, sr)
        features.append(feats)

    return features

def process_sample(example):
    audio = example['audio']['array']
    sr = example['audio']['sampling_rate']

    if len(audio) < MAX_SAMPLES:
        audio = fix_length(audio, size=MAX_SAMPLES)
    else:
        audio = audio[:MAX_SAMPLES]

    chunks = segment_and_extract(audio, sr)

    if not chunks:
        return {'input_values': None, 'label': None}

    # If your CNN expects fixed-size input (e.g., 7344), stack and pad to longest
    max_chunks = 5  # e.g., max 5 chunks (you can compute real max)
    chunk_dim = len(chunks[0])

    # Pad with zeros if fewer than max_chunks
    padded_chunks = np.zeros((max_chunks, chunk_dim))
    num_chunks = min(len(chunks), max_chunks)
    padded_chunks[:num_chunks] = chunks[:num_chunks]

    # Reshape to fit Conv1D: (timesteps, channels) = (chunk_dim, 1)
    final_input = padded_chunks.flatten()  # shape: (max_chunks * chunk_dim,)
    final_input = np.expand_dims(final_input, axis=-1)  # for Conv1D: (timesteps, 1)

    return {
        'input_values': final_input.astype(np.float32),
        'label': example['label']
    }


In [8]:
# Filter out samples with invalid audio if needed
dataset = dataset.filter(lambda x: x['audio'] is not None)

# Map and remove failed cases
processed = dataset["train"].map(process_sample, remove_columns=dataset["train"].column_names)
processed = processed.filter(lambda x: x['input_values'] is not None)


Filter:   0%|          | 0/40000 [00:00<?, ? examples/s]

Filter:   0%|          | 0/5000 [00:00<?, ? examples/s]

Map:   0%|          | 0/40000 [00:00<?, ? examples/s]

/usr/local/lib/python3.11/dist-packages/librosa/core/pitch.py:103: UserWarning: Trying to estimate tuning from empty frequency set.
  return pitch_tuning(


Filter:   0%|          | 0/40000 [00:00<?, ? examples/s]

In [15]:
len(processed[1]["input_values"])

330

In [ ]:
processed.save_to_disk("/kaggle/working/processed_dataset")


modeling

In [13]:
def build_cnn_model(input_shape=(7344, 1), num_classes=7):
    inputs = Input(shape=input_shape)

    x = Conv1D(128, kernel_size=3, padding='same', activation='relu')(inputs)
    x = BatchNormalization()(x)
    x = MaxPooling1D(pool_size=2)(x)

    x = Conv1D(128, kernel_size=3, padding='same', activation='relu')(x)
    x = BatchNormalization()(x)
    x = MaxPooling1D(pool_size=2)(x)
    x = Dropout(0.3)(x)

    x = Conv1D(64, kernel_size=3, padding='same', activation='relu')(x)
    x = BatchNormalization()(x)
    x = MaxPooling1D(pool_size=2)(x)

    x = Conv1D(64, kernel_size=3, padding='same', activation='relu')(x)
    x = BatchNormalization()(x)
    x = MaxPooling1D(pool_size=2)(x)
    x = Dropout(0.3)(x)

    x = Conv1D(32, kernel_size=3, padding='same', activation='relu')(x)
    x = BatchNormalization()(x)
    x = MaxPooling1D(pool_size=2)(x)
    x = Dropout(0.3)(x)

    x = Flatten()(x)
    x = Dense(128, activation='relu')(x)
    x = BatchNormalization()(x)
    outputs = Dense(num_classes, activation='softmax')(x)

    return Model(inputs, outputs, name="CNN_Model")


In [16]:
def build_cnn_bilstm_model(input_shape=(7344, 1), num_classes=7):
    inputs = Input(shape=input_shape)

    x = Conv1D(128, kernel_size=3, padding='same', activation='relu')(inputs)
    x = BatchNormalization()(x)
    x = MaxPooling1D(pool_size=2)(x)

    x = Conv1D(128, kernel_size=3, padding='same', activation='relu')(x)
    x = BatchNormalization()(x)
    x = MaxPooling1D(pool_size=2)(x)
    x = Dropout(0.3)(x)

    x = Conv1D(64, kernel_size=3, padding='same', activation='relu')(x)
    x = BatchNormalization()(x)
    x = MaxPooling1D(pool_size=2)(x)

    x = Conv1D(64, kernel_size=3, padding='same', activation='relu')(x)
    x = BatchNormalization()(x)
    x = MaxPooling1D(pool_size=2)(x)
    x = Dropout(0.3)(x)

    x = Bidirectional(LSTM(64, return_sequences=True))(x)

    x = Conv1D(32, kernel_size=3, padding='same', activation='relu')(x)
    x = BatchNormalization()(x)
    x = MaxPooling1D(pool_size=2)(x)
    x = Dropout(0.3)(x)

    x = Flatten()(x)
    x = Dense(128, activation='relu')(x)
    x = BatchNormalization()(x)
    outputs = Dense(num_classes, activation='softmax')(x)

    return Model(inputs, outputs, name="CNN_BiLSTM_Model")


In [17]:
def build_ensemble_model(input_shape=(7344, 1), num_classes=7):
    input_layer = Input(shape=input_shape)

    cnn_model = build_cnn_model(input_shape, num_classes)
    bilstm_model = build_cnn_bilstm_model(input_shape, num_classes)

    cnn_output = cnn_model(input_layer)
    bilstm_output = bilstm_model(input_layer)

    merged_output = Average()([cnn_output, bilstm_output])

    return Model(inputs=input_layer, outputs=merged_output, name="Ensemble_Model")


In [ ]:
model = build_ensemble_model()
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.summary()


In [ ]:
import pandas as pd
results = []


In [27]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, average_precision_score

import numpy as np

from tensorflow.keras.optimizers import Adam

def train_and_evaluate(model, model_name, X_train, y_train, X_val, y_val, X_test, y_test,
                       learning_rate=0.0001, batch_size=64, epochs=150):
    optimizer = Adam(learning_rate=learning_rate)
    model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])

    model.fit(X_train, y_train, batch_size=batch_size, epochs=epochs, validation_data=(X_val, y_val))

    y_pred = model.predict(X_test)
    y_pred_labels = np.argmax(y_pred, axis=1)
    y_true = np.argmax(y_test, axis=1)

    acc = accuracy_score(y_true, y_pred_labels)
    prec = precision_score(y_true, y_pred_labels, average='weighted')
    rec = recall_score(y_true, y_pred_labels, average='weighted')
    f1 = f1_score(y_true, y_pred_labels, average='weighted')

    try:
        auc_roc = roc_auc_score(y_test, y_pred, average='weighted', multi_class='ovr')
    except ValueError:
        auc_roc = None  # W razie problemu z AUC

    try:
        auc_prc = average_precision_score(y_test, y_pred, average='weighted')
    except ValueError:
        auc_prc = None

    return {
        'model_name': model_name,
        'accuracy': acc,
        'precision': prec,
        'recall': rec,
        'f1_score': f1,
        'auc_roc': auc_roc,
        'auc_prc': auc_prc
    }



In [19]:
processed

Dataset({
    features: ['label', 'input_values'],
    num_rows: 40000
})

In [ ]:
import numpy as np
from tensorflow.keras.utils import to_categorical

def prepare_data(data, num_classes):
    X = []
    y = []

    for example in data:
        input_values = example['input_values']
        label = example['label']

        X.append(input_values)
        y.append(label)

    # Padding/truncating 
    max_len = max(len(x) for x in X)
    X_padded = np.array([np.pad(x, (0, max_len - len(x))) for x in X])
    
    # Add chanell dimension (dla Conv1D: (samples, timesteps, channels))
    X_padded = np.expand_dims(X_padded, axis=-1)

    y_categorical = to_categorical(y, num_classes=num_classes)

    return X_padded, y_categorical, max_len


In [ ]:
from sklearn.model_selection import train_test_split


X_all, y_all, max_len = prepare_data(processed, num_classes=10)


input_shape = (max_len, 1)

X_train, X_temp, y_train_cat, y_temp_cat = train_test_split(X_all, y_all, test_size=0.3, random_state=42)
X_val, X_test, y_val_cat, y_test_cat = train_test_split(X_temp, y_temp_cat, test_size=0.5, random_state=42)


In [25]:
results = []

# CNN model
cnn_model = build_cnn_model(input_shape=input_shape, num_classes=10)
results.append(train_and_evaluate(cnn_model, "CNN_Model", X_train, y_train_cat, X_val, y_val_cat, X_test, y_test_cat))

# CNN+BiLSTM model
cnn_bilstm_model = build_cnn_bilstm_model(input_shape=input_shape, num_classes=10)
results.append(train_and_evaluate(cnn_bilstm_model, "CNN_BiLSTM_Model", X_train, y_train_cat, X_val, y_val_cat, X_test, y_test_cat))

# Ensemble model
ensemble_model = build_ensemble_model(input_shape=input_shape, num_classes=10)
results.append(train_and_evaluate(ensemble_model, "Ensemble_Model", X_train, y_train_cat, X_val, y_val_cat, X_test, y_test_cat))


I0000 00:00:1748888210.376141      35 gpu_device.cc:2022] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 13942 MB memory:  -> device: 0, name: Tesla T4, pci bus id: 0000:00:04.0, compute capability: 7.5
I0000 00:00:1748888210.376764      35 gpu_device.cc:2022] Created device /job:localhost/replica:0/task:0/device:GPU:1 with 13942 MB memory:  -> device: 1, name: Tesla T4, pci bus id: 0000:00:05.0, compute capability: 7.5


Epoch 1/50


I0000 00:00:1748888218.095327     115 service.cc:148] XLA service 0x7e128c0080a0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
I0000 00:00:1748888218.095756     115 service.cc:156]   StreamExecutor device (0): Tesla T4, Compute Capability 7.5
I0000 00:00:1748888218.095776     115 service.cc:156]   StreamExecutor device (1): Tesla T4, Compute Capability 7.5
I0000 00:00:1748888218.756837     115 cuda_dnn.cc:529] Loaded cuDNN version 90300
2025-06-02 18:17:01.263631: E external/local_xla/xla/service/slow_operation_alarm.cc:65] Trying algorithm eng3{k11=0} for conv (f32[128,128,1,3]{3,2,1,0}, u8[0]{0}) custom-call(f32[128,64,1,165]{3,2,1,0}, f32[128,64,1,165]{3,2,1,0}), window={size=1x165 pad=0_0x1_1}, dim_labels=bf01_oi01->bf01, custom_call_target="__cudnn$convForward", backend_config={"cudnn_conv_backend_config":{"activation_mode":"kNone","conv_result_scale":1,"leakyrelu_alpha":0,"side_input_scale":0},"force_earliest_schedule":false,"operation_qu

 16/438 ━━━━━━━━━━━━━━━━━━━━ 4s 11ms/step - accuracy: 0.0953 - loss: 2.9930  

I0000 00:00:1748888225.214446     115 device_compiler.h:188] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.


438/438 ━━━━━━━━━━━━━━━━━━━━ 24s 24ms/step - accuracy: 0.1171 - loss: 2.8182 - val_accuracy: 0.2175 - val_loss: 2.1831
Epoch 2/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.1898 - loss: 2.3554 - val_accuracy: 0.3100 - val_loss: 2.0050
Epoch 3/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.2649 - loss: 2.1106 - val_accuracy: 0.3393 - val_loss: 1.8830
Epoch 4/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.3120 - loss: 1.9557 - val_accuracy: 0.3483 - val_loss: 1.8185
Epoch 5/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.3335 - loss: 1.8643 - val_accuracy: 0.3607 - val_loss: 1.7718
Epoch 6/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.3438 - loss: 1.8191 - val_accuracy: 0.3527 - val_loss: 1.7708
Epoch 7/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.3472 - loss: 1.7951 - val_accuracy: 0.3688 - val_loss: 1.7445
Epoch 8/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.3592 - loss: 1.7774 - val_accuracy: 0.3595 - va

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Epoch 1/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 20s 20ms/step - accuracy: 0.1159 - loss: 2.8371 - val_accuracy: 0.2512 - val_loss: 2.2085
Epoch 2/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 8s 18ms/step - accuracy: 0.2025 - loss: 2.3024 - val_accuracy: 0.2928 - val_loss: 2.0989
Epoch 3/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 8s 18ms/step - accuracy: 0.2897 - loss: 2.0628 - val_accuracy: 0.3222 - val_loss: 1.9286
Epoch 4/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 8s 18ms/step - accuracy: 0.3302 - loss: 1.9190 - val_accuracy: 0.3365 - val_loss: 1.8314
Epoch 5/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 8s 18ms/step - accuracy: 0.3446 - loss: 1.8481 - val_accuracy: 0.3350 - val_loss: 1.8293
Epoch 6/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 8s 18ms/step - accuracy: 0.3572 - loss: 1.8022 - val_accuracy: 0.3428 - val_loss: 1.7948
Epoch 7/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 8s 18ms/step - accuracy: 0.3575 - loss: 1.7856 - val_accuracy: 0.3455 - val_loss: 1.7759
Epoch 8/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 8s 18ms/step - accuracy: 0.3634 - loss: 1.7754 - val_acc

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Epoch 1/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 31s 33ms/step - accuracy: 0.1857 - loss: 2.4472 - val_accuracy: 0.3595 - val_loss: 1.8401
Epoch 2/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 13s 29ms/step - accuracy: 0.3327 - loss: 1.9782 - val_accuracy: 0.3597 - val_loss: 1.8227
Epoch 3/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 13s 29ms/step - accuracy: 0.3411 - loss: 1.9003 - val_accuracy: 0.3657 - val_loss: 1.7880
Epoch 4/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 13s 30ms/step - accuracy: 0.3419 - loss: 1.8627 - val_accuracy: 0.3643 - val_loss: 1.7766
Epoch 5/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 13s 29ms/step - accuracy: 0.3453 - loss: 1.8379 - val_accuracy: 0.3630 - val_loss: 1.7700
Epoch 6/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 13s 30ms/step - accuracy: 0.3510 - loss: 1.8138 - val_accuracy: 0.3493 - val_loss: 1.7958
Epoch 7/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 13s 29ms/step - accuracy: 0.3602 - loss: 1.7811 - val_accuracy: 0.3627 - val_loss: 1.7542
Epoch 8/50
438/438 ━━━━━━━━━━━━━━━━━━━━ 13s 29ms/step - accuracy: 0.3579 - loss: 1.7883 - 

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In the context of all modeliemy y some analogies:
Initial stages:
Epoch 1: accuracy: 11.7%, val_accuracy: 21.8% - random model (with 7 classes random selection gives ~14% hits), so the beginning was difficult.
By the era of 5/10 there was a meaningful increase in val_accuracy (to ~35-39%) - the model begins to learn the relevant features of the signal, but still far from satisfactory performance.
Stabilization and further improvement:
From around epoch 10, accuracy on the val_accuracy set gradually improves and stabilizes within 37-39% and does not make significant improvements.
At the same time, training accuracy also increases, but without much excessive advantage over validation, suggesting moderate overfitting (Dropout and BatchNorm layers work well).

We will also try to train on higher amount of epochs and check if models will get better.

In [28]:
results = []

# CNN model
cnn_model = build_cnn_model(input_shape=input_shape, num_classes=10)
results.append(train_and_evaluate(cnn_model, "CNN_Model", X_train, y_train_cat, X_val, y_val_cat, X_test, y_test_cat))

# CNN+BiLSTM model
cnn_bilstm_model = build_cnn_bilstm_model(input_shape=input_shape, num_classes=10)
results.append(train_and_evaluate(cnn_bilstm_model, "CNN_BiLSTM_Model", X_train, y_train_cat, X_val, y_val_cat, X_test, y_test_cat))

# Ensemble model
ensemble_model = build_ensemble_model(input_shape=input_shape, num_classes=10)
results.append(train_and_evaluate(ensemble_model, "Ensemble_Model", X_train, y_train_cat, X_val, y_val_cat, X_test, y_test_cat))


Epoch 1/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 20s 22ms/step - accuracy: 0.1163 - loss: 2.8181 - val_accuracy: 0.3485 - val_loss: 2.0530
Epoch 2/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 4s 8ms/step - accuracy: 0.2001 - loss: 2.3232 - val_accuracy: 0.3483 - val_loss: 1.9361
Epoch 3/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.2707 - loss: 2.0910 - val_accuracy: 0.3465 - val_loss: 1.8767
Epoch 4/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 4s 8ms/step - accuracy: 0.3164 - loss: 1.9499 - val_accuracy: 0.3652 - val_loss: 1.7910
Epoch 5/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 4s 8ms/step - accuracy: 0.3385 - loss: 1.8477 - val_accuracy: 0.3618 - val_loss: 1.7767
Epoch 6/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 4s 8ms/step - accuracy: 0.3425 - loss: 1.8172 - val_accuracy: 0.3627 - val_loss: 1.7629
Epoch 7/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 4s 8ms/step - accuracy: 0.3503 - loss: 1.7940 - val_accuracy: 0.3623 - val_loss: 1.7614
Epoch 8/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.3569 - loss: 1.7840 - val_ac

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Epoch 1/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 18s 20ms/step - accuracy: 0.1244 - loss: 2.8167 - val_accuracy: 0.2697 - val_loss: 2.1488
Epoch 2/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 8s 18ms/step - accuracy: 0.2020 - loss: 2.3138 - val_accuracy: 0.3333 - val_loss: 2.0171
Epoch 3/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 8s 19ms/step - accuracy: 0.2788 - loss: 2.0659 - val_accuracy: 0.3440 - val_loss: 1.9097
Epoch 4/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 8s 18ms/step - accuracy: 0.3221 - loss: 1.9290 - val_accuracy: 0.3590 - val_loss: 1.8081
Epoch 5/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 8s 18ms/step - accuracy: 0.3414 - loss: 1.8513 - val_accuracy: 0.3598 - val_loss: 1.7771
Epoch 6/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 8s 18ms/step - accuracy: 0.3520 - loss: 1.8057 - val_accuracy: 0.3578 - val_loss: 1.7693
Epoch 7/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 8s 19ms/step - accuracy: 0.3597 - loss: 1.7807 - val_accuracy: 0.3610 - val_loss: 1.7642
Epoch 8/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 8s 18ms/step - accuracy: 0.3515 - loss: 1.7869 -

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Epoch 1/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 29s 32ms/step - accuracy: 0.1694 - loss: 2.4780 - val_accuracy: 0.3568 - val_loss: 1.8693
Epoch 2/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 13s 30ms/step - accuracy: 0.3180 - loss: 2.0061 - val_accuracy: 0.3567 - val_loss: 1.8280
Epoch 3/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 13s 29ms/step - accuracy: 0.3403 - loss: 1.9123 - val_accuracy: 0.3583 - val_loss: 1.7898
Epoch 4/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 13s 29ms/step - accuracy: 0.3468 - loss: 1.8627 - val_accuracy: 0.3562 - val_loss: 1.7755
Epoch 5/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 13s 30ms/step - accuracy: 0.3456 - loss: 1.8395 - val_accuracy: 0.3627 - val_loss: 1.7653
Epoch 6/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 13s 29ms/step - accuracy: 0.3583 - loss: 1.8033 - val_accuracy: 0.3665 - val_loss: 1.7563
Epoch 7/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 13s 30ms/step - accuracy: 0.3542 - loss: 1.8041 - val_accuracy: 0.3677 - val_loss: 1.7473
Epoch 8/150
438/438 ━━━━━━━━━━━━━━━━━━━━ 13s 29ms/step - accuracy: 0.3533 - loss: 1

Average results are a bit better (around 43%) but it is not significant difference. Here the problem is definitely in the amount of characteristics extraced for each sample. 